qubes/tests/integ/dispvm.py

```diff
@@ -221,6 +225,118 @@ def test_010_simple_dvm_run(self):
         finally:
             self.loop.run_until_complete(dispvm.cleanup())
 
+    def test_011_dvm_run_preload_invalid_max(self):
+        # TODO: couldn't make the assert raise on unit tests.
+        cases_invalid = ["a", "-1", "1 1"]
+        for value in cases_invalid:
+            with self.subTest(value=value):
+                with self.assertRaises(qubes.exc.QubesValueError):
+                    self.appvm.features["preload-dispvm-max"] = value
+
+    def test_011_dvm_run_preload_invalid_list(self):
+        # TODO: couldn't make the assert raise on unit tests.
+        dispvm = self.loop.run_until_complete(
+            qubes.vm.dispvm.DispVM.from_appvm(self.disp_base)
+        )
+        cases_invalid = [
+            "notaqube",  # not a qube
+            f"{self.disp_base}",  # not derived from wanted appvm
+            f"{dispvm.name} {dispvm.name}",  # duplicate
+        ]
+        try:
+            with self.assertRaises(qubes.exc.QubesValueError):
+                # exceeds the limit (0 if unset)
+                self.appvm.features["preload-dispvm"] = f"{dispvm.name}"
+            self.appvm.features["preload-dispvm-max"] = "2"
+            for value in cases_invalid:
+                with self.subTest(value=value):
+                    with self.assertRaises(qubes.exc.QubesValueError):
+                        self.appvm.features["preload-dispvm"] = value
+        finally:
+            self.loop.run_until_complete(dispvm.cleanup())
+
+    def test_011_dvm_run_preload_reject_max(self):
+        with self.assertRaises(qubes.exc.QubesException):
+            self.loop.run_until_complete(
+                qubes.vm.dispvm.DispVM.from_appvm(self.disp_base, preload=True)
+            )
+
+    def test_011_dvm_run_preload(self):
+        self.disp_base.features["preload-dispvm-max"] = "1"
+        dispvm = self.loop.run_until_complete(
+            qubes.vm.dispvm.DispVM.from_appvm(self.disp_base, preload=True)
+        )
+        self.disp_base.features["preload-dispvm-max"] = False
+        try:
+            self.assertEqual(self.disp_base.get_feat_preload(), [dispvm.name])
+            self.assertTrue(dispvm.is_preloaded())
+            self.assertTrue(dispvm.features.get("internal", False))
+            self.assertTrue(dispvm.is_paused())
+            # TODO: how to assert call was made to the desired qube?
+            # I saw qubes-core-admin-client/qubesadmin/tests/tools/qvm_run.py
+            # but it has extra things in tests/__init__.py (expected_calls and
+            # service_calls_).
+            ret = qvm_run.main(self.qvm_run_preload_args, app=self.app)
+            self.assertEqual(ret, 0)
+            self.assertEqual(self.disp_base.get_feat_preload(), [])
+            # TODO: these tests are too late, when qvm_run returns, the qube
```

Comment:
If you change the call to use not qubes.GetDate, but something that waits for input (even qubes.VMShell), then you can have `qvm-run -p` started (via asyncio, as noted above), check what you need and only then send EOF to finalize the service call.
Note that starting the process doesn't mean you instantly get the dispvm, but if you go with the hostname idea, you can wait for the service to print it, then do checks, and then send EOF.